docs/fundamentals/ruler.mdx

```diff
@@ -0,0 +1,279 @@
+---
+title: "RULER"
+icon: "ruler"
+description: "Learn how to use RULER to automatically reward your agents."
+---
+
+# 📏RULER: Relative Universal LLM-Elicited Rewards
+
+RULER (Relative Universal LLM-Elicited Rewards) is a general-purpose reward function that uses an LLM-as-judge to rank multiple agent trajectories. It requires no labeled data, expert feedback, or hand-crafted reward functions, yet reliably improves agent performance.
+
+<div align="center">
+  <img src="/images/ruler-results.png" alt="RULER Performance Results" style={{maxWidth: "100%", height: "auto"}} />
+  <p><em>RULER performance across multiple tasks at launch. In 3 out of 4 tasks, models trained with RULER slightly outperform those trained with hand-crafted reward functions. See the full <a href="https://openpipe.ai/blog/ruler">launch announcement</a> for details.</em></p>
+</div>
+
+## Key Benefits
+
+- **No labeled data required**: RULER works by comparing trajectories against each other
+- **General-purpose**: Can be applied to a wide variety of RL tasks without modification
+- **Fast development**: Can reduce implementation time by 2-3x compared to hand-crafted rewards
+- **Strong performance**: Often matches or exceeds hand-crafted reward functions
+
+## How RULER Works
+
+RULER leverages two key insights:
+
+1. **Relative scoring is easier than absolute scoring**: It's easier for an LLM to rank several solutions relative to each other than to score them in isolation
+2. **GRPO only needs relative scores**: Since GRPO normalizes scores within each group, only the relative rankings matter, not absolute values
+
+The process:
+1. Generate N trajectories for a given scenario
+2. Pass all N trajectories to RULER
+3. RULER deduplicates common prefixes (e.g., identical system messages)
+4. An LLM judge scores each trajectory from 0 to 1 based on goal achievement
+5. These scores are used directly as rewards in GRPO training
+
+## Basic Usage
+
+```python
+import art
+from art.rewards import ruler_score_group
+
+# Create a TrajectoryGroup from your trajectories
+group = art.TrajectoryGroup([...])  # List of art.Trajectory objects
+
+# Use RULER to score them
+judged_group = await ruler_score_group(
+    group,
+    {"model": "openai/o3"},
```

Comment:
Why nest model in a dictionary, instead of making it a top-level parameter?

Reply:
lets us customize other litellm params, which is useful for eg. using a local model with a specific api endpoint.

Reply:
that said, `model` is frequently going to be the only thing you need to customize. I'll make that its own explicit argument and have a separate one for additional litellm params.